rust/otap-dataflow/crates/otap/src/otap_grpc/otlp/server.rs

```diff
@@ -8,24 +8,74 @@
 //! requires it
 
 use std::convert::Infallible;
+use std::sync::{Arc, Mutex};
 use std::task::Poll;
 
+use crate::accessory::slots::{Config as SlotsConfig, SlotKey, State as SlotsState};
 use crate::pdata::{OtapPdata, OtlpProtoBytes};
 use crate::proto::opentelemetry::collector::logs::v1::ExportLogsServiceResponse;
 use crate::proto::opentelemetry::collector::metrics::v1::ExportMetricsServiceResponse;
 use crate::proto::opentelemetry::collector::trace::v1::ExportTraceServiceResponse;
 use futures::future::BoxFuture;
 use http::{Request, Response};
+use otap_df_engine::control::{AckMsg, CallData, NackMsg};
 use otap_df_engine::shared::receiver::EffectHandler;
+use otap_df_engine::{Interests, ProducerEffectHandlerExtension};
 use prost::Message;
 use prost::bytes::Buf;
+use tokio::sync::oneshot;
 use tonic::Status;
 use tonic::body::Body;
 use tonic::codec::{
     Codec, CompressionEncoding, DecodeBuf, Decoder, EnabledCompressionEncodings, EncodeBuf, Encoder,
 };
 use tonic::server::{Grpc, NamedService, UnaryService};
 
+/// Shared state for correlating requests with responses using slots
+pub type SharedCorrelationState =
+    Arc<Mutex<SlotsState<oneshot::Sender<Result<(), NackMsg<OtapPdata>>>>>>;
+
+/// Route an Ack response back to the appropriate correlation slot
+pub fn route_ack_response(state: &SharedCorrelationState, ack: AckMsg<OtapPdata>) {
+    route_response(state, ack.calldata, Ok(()));
+}
+
+/// Route a Nack response back to the appropriate correlation slot
+pub fn route_nack_response(state: &SharedCorrelationState, mut nack: NackMsg<OtapPdata>) {
+    let calldata = std::mem::take(&mut nack.calldata);
+    route_response(state, calldata, Err(nack));
+}
+
+/// Internal helper to route responses to slots
+fn route_response(
+    state: &SharedCorrelationState,
+    calldata: CallData,
+    result: Result<(), NackMsg<OtapPdata>>,
+) {
+    // Decode slot key from calldata
+    let slot_key: SlotKey = match calldata.try_into() {
+        Ok(data) => data,
+        Err(_) => {
+            // Invalid calldata format
+            return;
+        }
+    };
+
+    // Try to get the channel from the slot
+    if let Ok(mut state) = state.lock() {
+        if let Some(sender) = state.get_if_current(slot_key) {
+            match sender.send(result) {
+                Ok(()) => {
+                    // Sent
+                }
+                Err(_) => {
+                    // E.g., channel closed
+                }
+            }
```

Comment:
Can't we put this part outside of the lock to minimize the time the mutex is locked?

Reply:
For my own understanding, does it matter? Do we expect any contention for the `Mutex` here given the thread-per-core setup?

Reply:
Done! I made this more functional, also improved another case where the lock was taken and unwrapped less gracefully.